File: test-junit5/src/test/java/io/micronaut/test/junit5/PropertySourceMapTest.java

package io.micronaut.test.junit5;

import io.micronaut.context.annotation.Property;
import io.micronaut.core.util.CollectionUtils;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.*;

import javax.annotation.Nonnull;
import java.util.Map;

@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class PropertySourceMapTest implements TestPropertyProvider {

    @Property(name = "foo.bar")
    String val;

    @Test
    void testPropertySource() {
        Assertions.assertEquals("one", val);
    }

    @Nonnull
    @Override
    public Map<String, String> getProperties() {
        return CollectionUtils.mapOf(
                "foo.bar", "one",
                "foo.baz", "two"
        );
    }
}

File: test-suite-python/src/test/python/micronaut/test/python/PropertySourceMapTest.py
from typing import Annotated

from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from micronaut.test.support import TestPropertyProvider
from org.junit.jupiter.api import Disabled, Test, TestInstance


# TODO(python): TestPropertyProvider.getProperties() is called before the Micronaut context (and with it the
# GraalPy runtime) exists, so a Python test cannot provide properties this way ("GraalPy context has not been initialized").
@Disabled("TODO(python): TestPropertyProvider needs the GraalPy runtime before the context starts")
@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class PropertySourceMapTest(TestPropertyProvider):

    val: Annotated[str, Property(name="foo.bar")]

    @Test
    def test_property_source(self):
        assert "one" == self.val

    def getProperties(self) -> dict[str, str]:
        return {
            "foo.bar": "one",
            "foo.baz": "two",
        }
